Index: package/network/services/dnsmasq/Makefile
===================================================================
--- package/network/services/dnsmasq/Makefile	(revision 46893)
+++ package/network/services/dnsmasq/Makefile	(working copy)
@@ -116,7 +116,7 @@
 		$(if $(CONFIG_PACKAGE_dnsmasq_$(BUILD_VARIANT)_ipset),,-DNO_IPSET)
 	COPTS += $(if $(CONFIG_LIBNETTLE_MINI),-DNO_GMP,)
 else
-	COPTS += -DNO_AUTH -DNO_IPSET
+	COPTS += -DNO_AUTH
 endif
 
 MAKE_FLAGS := \
