Index: target/linux/ramips/image/Makefile
===================================================================
--- target/linux/ramips/image/Makefile	(revision 43594)
+++ target/linux/ramips/image/Makefile	(working copy)
@@ -781,6 +781,7 @@
 na930_mtd_size=20971520
 Image/Build/Profile/NA930=$(call BuildFirmware/CustomFlash/$(1),$(1),na930,NA930,$(na930_mtd_size))
 Image/Build/Profile/MZK-750DHP=$(call BuildFirmware/Default8M/$(1),$(1),mzk-750dhp,MZK-750DHP)
+Image/Build/Profile/HiWiFi-HC5761=$(call BuildFirmware/Default16M/$(1),$(1),hiwifi-hc5761,HiWiFi-HC5761)
 Image/Build/Profile/Y1=$(call BuildFirmware/Default16M/$(1),$(1),Lenovo-y1,Y1)
 Image/Build/Profile/Y1S=$(call BuildFirmware/Default16M/$(1),$(1),Lenovo-y1s,Y1S)
 Image/Build/Profile/MLW221=$(call BuildFirmware/Default16M/$(1),$(1),mlw221,MLW221)
@@ -806,6 +807,7 @@
 	$(call Image/Build/Profile/WHR1166D,$(1))
 	$(call Image/Build/Profile/MZK-750DHP,$(1))
 	$(call Image/Build/Profile/NA930,$(1))
+	$(call Image/Build/Profile/HiWiFi-HC5761,$(1))
 	$(call Image/Build/Profile/Y1,$(1))
 	$(call Image/Build/Profile/Y1S,$(1))
 	$(call Image/Build/Profile/MLW221,$(1))
Index: target/linux/ramips/dts/HiWiFi-HC5761.dts
===================================================================
--- target/linux/ramips/dts/HiWiFi-HC5761.dts	(revision 0)
+++ target/linux/ramips/dts/HiWiFi-HC5761.dts	(revision 0)
@@ -0,0 +1,163 @@
+/dts-v1/;
+
+/include/ "mt7620a.dtsi"
+
+/ {
+	compatible = "HiWiFi-HC5761", "ralink,mt7620a-soc";
+	model = "HiWiFi HC5761";
+
+	chosen {
+		bootargs = "console=ttyS0,115200";
+	};
+
+	palmbus@10000000 {
+		sysc@0 {
+			ralink,gpiomux = "i2c", "jtag";
+			ralink,uartmux = "gpio";
+			ralink,wdtmux = <1>;
+		};
+		gpio0: gpio@600 {
+			status = "okay";
+		};
+		gpio2: gpio@660 {
+			status = "okay";
+		};
+		gpio3: gpio@688 {
+			status = "okay";
+		};
+
+		spi@b00 {
+			status = "okay";
+
+			m25p80@0 {
+				#address-cells = <1>;
+				#size-cells = <1>;
+				compatible = "w25q128";
+				reg = <0 0>;
+				linux,modalias = "m25p80", "w25q128";
+				spi-max-frequency = <10000000>;
+
+				partition@0 {
+					label = "u-boot";
+					reg = <0x0 0x30000>;
+					read-only;
+				};
+
+				partition@30000 {
+					label = "u-boot-env";
+					reg = <0x30000 0x10000>;
+					read-only;
+				};
+
+				factory: partition@40000 {
+					label = "factory";
+					reg = <0x40000 0x10000>;
+					read-only;
+				};
+
+				partition@50000 {
+					label = "firmware";
+					reg = <0x50000 0xf80000>;
+				};
+
+				partition@fd0000 {
+					label = "hwf_config";
+					reg = <0xfd0000 0x10000>;
+				};
+
+				partition@fe0000 {
+					label = "bdinfo";
+					reg = <0xfe0000 0x10000>;
+				};
+
+				partition@ff0000 {
+					label = "backup";
+					reg = <0xff0000 0x10000>;
+				};
+			};
+		};
+	};
+
+	ehci@101c0000 {
+		status = "okay";
+	};
+
+	ohci@101c1000 {
+		status = "okay";
+	};
+
+	sdhci@10130000 {
+		status = "okay";
+	};
+
+	pcie@10140000 {
+		status = "okay";
+	};
+
+	wmac@10180000 {
+		ralink,mtd-eeprom = <&factory 0>;
+	};
+
+	pinctrl {
+		state_default: pinctrl0 {
+			gpio {
+				ralink,group = "uartf", "wled", "nd_sd";
+				ralink,function = "gpio";
+			};
+			pa {
+				ralink,group = "pa";
+				ralink,function = "pa";
+			};
+		};
+	};
+
+	ethernet@10100000 {
+		pinctrl-names = "default";
+		pinctrl-0 = <&ephy_pins>;
+		mtd-mac-address = <&factory 0x4>;
+		ralink,port-map = "llllw";
+	};
+	
+	gpio-keys-polled {
+		compatible = "gpio-keys-polled";
+		#address-cells = <1>;
+		#size-cells = <0>;
+		poll-interval = <20>;
+		reset {
+			label = "reset";
+			gpios = <&gpio0 12 1>;
+			linux,code = <0x198>;
+		};
+	};
+
+	gpio-leds {
+		compatible = "gpio-leds";
+		system {
+			label = "HC5761:green:system";
+			gpios = <&gpio0 9 1>;
+		};
+		internet {
+			label = "HC5761:green:internet";
+			gpios = <&gpio0 11 1>;
+		};
+		wlan2p4 {
+			label = "HC5761:green:wlan-2p4";
+			gpios = <&gpio3 0 1>;
+		};
+		wlan5p {
+			label = "HC5761:green:wlan-5p";
+			gpios = <&gpio0 7 1>;
+		};
+	};
+
+	gpio_export {
+		compatible = "gpio-export";
+		#size-cells = <0>;
+		usbpower {
+			gpio-export,name = "usbpower";
+			gpio-export,output = <1>;
+			gpios = <&gpio0 13 0>;
+		};
+	};
+
+};
Index: target/linux/ramips/mt7620/profiles/HiWiFi-HC5761.mk
===================================================================
--- target/linux/ramips/mt7620/profiles/HiWiFi-HC5761.mk	(revision 0)
+++ target/linux/ramips/mt7620/profiles/HiWiFi-HC5761.mk	(revision 0)
@@ -0,0 +1,20 @@
+#
+# Copyright (C) 2011 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+
+define Profile/HiWiFi-HC5761
+ NAME:=HiWiFi HC5761
+ PACKAGES:=\
+   kmod-usb-core kmod-usb-dwc2 kmod-usb2 kmod-usb-storage  \
+   kmod-mmc-spi kmod-sdhci kmod-sdhci-mt7620 \
+   block-mount mountd kmod-fs-ext4 ppp-mod-pppoe e2fsprogs \
+   kmod-ledtrig-usbdev fdisk
+endef
+
+define Profile/HiWiFi-HC5761/Description
+ Default on SD cards and USB storage support.
+endef
+$(eval $(call Profile,HiWiFi-HC5761))
Index: target/linux/ramips/base-files/lib/ramips.sh
===================================================================
--- target/linux/ramips/base-files/lib/ramips.sh	(revision 43594)
+++ target/linux/ramips/base-files/lib/ramips.sh	(working copy)
@@ -355,6 +355,9 @@
 	*"ZBT WR8305RT")
 		name="wr8305rt"
 		;;
+	*"HiWiFi HC5761")
+		name="hiwifi-hc5761"
+		;;
 	*"Lenovo Y1")
 		name="y1"
 		;;
Index: target/linux/ramips/base-files/lib/upgrade/platform.sh
===================================================================
--- target/linux/ramips/base-files/lib/upgrade/platform.sh	(revision 43594)
+++ target/linux/ramips/base-files/lib/upgrade/platform.sh	(working copy)
@@ -106,6 +106,7 @@
 	wt3020 |\
 	x5 |\
 	x8 |\
+	hiwifi-hc5761 |\
 	y1 |\
 	y1s |\
 	zbt-wa05)
Index: target/linux/ramips/base-files/etc/board.d/01_leds
===================================================================
--- target/linux/ramips/base-files/etc/board.d/01_leds	(revision 43594)
+++ target/linux/ramips/base-files/etc/board.d/01_leds	(working copy)
@@ -212,6 +212,11 @@
 		set_usb_led "wr8305rt:usb"
 		set_wifi_led "wr8305rt:wifi"
 		;;
+	hiwifi-hc5761)
+		ucidef_set_led_default "system" "system" "HC5761:green:system" "1"
+		ucidef_set_led_netdev "internet" "internet" "HC5761:green:internet" "eth0.2"
+		set_wifi_led "HC5761:green:wlan-2p4"
+		;;
 	wt1520)
 		set_wifi_led "rt2800pci-phy0::radio"
 		;;
Index: target/linux/ramips/base-files/etc/board.d/02_network
===================================================================
--- target/linux/ramips/base-files/etc/board.d/02_network	(revision 43594)
+++ target/linux/ramips/base-files/etc/board.d/02_network	(working copy)
@@ -213,6 +213,7 @@
 		;;
 
 	y1 |\
+	hiwifi-hc5761 |\
 	y1s)
 		ucidef_set_interfaces_lan_wan "eth0.1" "eth0.2"
 		ucidef_add_switch "switch0" "1" "1"
@@ -296,6 +297,7 @@
 	psr-680w |\
 	sl-r7205 |\
 	y1 |\
+	hiwifi-hc5761 |\
 	y1s)
 		lan_mac=$(cat /sys/class/net/eth0/address)
 		lan_mac=$(macaddr_setbit_la "$lan_mac")
Index: target/linux/ramips/base-files/etc/diag.sh
===================================================================
--- target/linux/ramips/base-files/etc/diag.sh	(revision 43594)
+++ target/linux/ramips/base-files/etc/diag.sh	(working copy)
@@ -207,6 +207,9 @@
 	na930)
 		status_led="na930:blue:power"
 		;;
+  hiwifi-hc5761)
+    status_led="HC5761:green:system"
+    ;;
 	y1 | \
 	y1s)
 		status_led="lenovo:blue:power"
Index: target/linux/ramips/patches-3.14/0105-mtd-gd25q.patch
===================================================================
--- target/linux/ramips/patches-3.14/0105-mtd-gd25q.patch	(revision 0)
+++ target/linux/ramips/patches-3.14/0105-mtd-gd25q.patch	(revision 0)
@@ -0,0 +1,11 @@
+--- a/drivers/mtd/devices/m25p80.c
++++ b/drivers/mtd/devices/m25p80.c
+@@ -1162,6 +1162,8 @@ static const struct spi_device_id m25p_ids[] = {
+ 	{ "w25q128", INFO(0xef4018, 0, 64 * 1024, 256, SECT_4K) },
+ 	{ "w25q256", INFO(0xef4019, 0, 64 * 1024, 512, SECT_4K) },
+ 
++	{ "gd25q128b", INFO(0xc84018, 0, 64 * 1024, 256, SECT_4K)},
++
+ 	/* Catalyst / On Semiconductor -- non-JEDEC */
+ 	{ "cat25c11", CAT25_INFO(  16, 8, 16, 1, M25P_NO_ERASE | M25P_NO_FR) },
+ 	{ "cat25c03", CAT25_INFO(  32, 8, 16, 2, M25P_NO_ERASE | M25P_NO_FR) },
