Index: target/linux/ramips/base-files/lib/ramips.sh
===================================================================
--- target/linux/ramips/base-files/lib/ramips.sh	(revision 43770)
+++ target/linux/ramips/base-files/lib/ramips.sh	(working copy)
@@ -325,6 +325,9 @@
 	*"ZBT WR8305RT")
 		name="wr8305rt"
 		;;
+	*"HiWiFi HC5761")
+		name="hiwifi-hc5761"
+		;;
 	*"Lenovo Y1")
 		name="y1"
 		;;
Index: target/linux/ramips/base-files/lib/upgrade/platform.sh
===================================================================
--- target/linux/ramips/base-files/lib/upgrade/platform.sh	(revision 43770)
+++ target/linux/ramips/base-files/lib/upgrade/platform.sh	(working copy)
@@ -98,6 +98,7 @@
 	x5 |\
 	x8 |\
 	zbt-wa05 |\
+	hiwifi-hc5761 |\
 	y1 |\
 	y1s)
 		[ "$magic" != "27051956" ] && {
Index: target/linux/ramips/base-files/etc/uci-defaults/01_leds
===================================================================
--- target/linux/ramips/base-files/etc/uci-defaults/01_leds	(revision 43770)
+++ target/linux/ramips/base-files/etc/uci-defaults/01_leds	(working copy)
@@ -203,6 +203,11 @@
 		set_usb_led "wr8305rt:usb"
 		set_wifi_led "wr8305rt:wifi"
 		;;
+	hiwifi-hc5761)
+		ucidef_set_led_default "system" "system" "HC5761:green:system" "1"
+		ucidef_set_led_netdev "internet" "internet" "HC5761:green:internet" "eth0.2"
+		set_wifi_led "HC5761:green:wlan-2p4"
+		;;
 	y1 |\
 	y1s)
 		ucidef_set_led_default "power" "power" "lenovo:blue:power" "1"
Index: target/linux/ramips/base-files/etc/uci-defaults/02_network
===================================================================
--- target/linux/ramips/base-files/etc/uci-defaults/02_network	(revision 43770)
+++ target/linux/ramips/base-files/etc/uci-defaults/02_network	(working copy)
@@ -183,6 +183,7 @@
 		ucidef_add_switch_vlan "switch0" "2" "4 6t"
 		;;
 
+	hiwifi-hc5761 |\
 	y1 |\
 	y1s)
 		ucidef_set_interfaces_lan_wan "eth0.1" "eth0.2"
Index: target/linux/ramips/base-files/etc/diag.sh
===================================================================
--- target/linux/ramips/base-files/etc/diag.sh	(revision 43770)
+++ target/linux/ramips/base-files/etc/diag.sh	(working copy)
@@ -188,6 +188,9 @@
 	na930)
 		status_led="na930:blue:power"
 		;;
+	hiwifi-hc5761)
+		status_led="HC5761:green:system"
+		;;
 	y1 | \
 	y1s)
 		status_led="lenovo:blue:power"
